File: src/main/java/unyo/runtime/LMNtalRuntime.java
package unyo.runtime;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;

import java.util.ArrayList;
import java.util.List;

import unyo.entity.Graphy;
import unyo.entity.Graph;

public class LMNtalRuntime {

    private static List<String> readStdout(Process process) {
        final List<String> ret = new ArrayList<String>();
        final Process p = process;

        Thread inThread = new Thread() {
            @Override
            public void run() {
                try {
                    BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()));
                    String buf;
                    while ((buf = br.readLine()) != null) {
                        ret.add(buf);
                    }
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        };

        Thread outThread = new Thread() {
            @Override
            public void run() {
                PrintWriter pw = new PrintWriter(new OutputStreamWriter(p.getOutputStream()));
                while (!pw.checkError()) {
                    pw.println("");
                }
                pw.close();
            }
        };

        inThread.start();
        outThread.start();

        try {
            inThread.join();
            outThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        return ret;
    }

    public static Graph execute(File file, List<String> options) {

        try {
            ProcessBuilder pb = new ProcessBuilder(java.util.Arrays.asList("/Users/charlie/Documents/slim/slim/src/slim", "--json-dump", file.getAbsolutePath()));
            pb.redirectErrorStream(true);

            Process p = pb.start();
            List<String> lines = readStdout(p);

            return Graphy.fromString(lines.get(0));
        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }

}

File: src/main/scala/unyo/runtime/LMNtalRuntime.scala
package unyo.runtime

import unyo.entity.{Graph, Graphy}

import java.io.{BufferedReader,InputStreamReader,PrintWriter,OutputStreamWriter}
import java.io.{File,IOException}
import collection.mutable.Buffer
import collection.JavaConversions._

import scala.actors.Actor._

class LMNtalRuntime(file: File, options: java.util.List[String]) {

  val pb = new ProcessBuilder(Buffer("/Users/charlie/Documents/slim/slim/src/slim", "--json-dump") ++ options ++ Buffer(file.getAbsolutePath))
  pb.redirectErrorStream(true)
  val p = pb.start

  val reader = scala.actors.Actor.actor {
    val br = new BufferedReader(new InputStreamReader(p.getInputStream))
    loop {
      react {
        case "next" => reply(br.readLine)
        case "end" => exit
      }
    }
  }

  val writer = actor {
    val pw = new PrintWriter(new OutputStreamWriter(p.getOutputStream))
    loop {
      react {
        case "next" => pw.println("")
        case "end" => exit
      }
    }
  }

  var finished = false
  var _next: Option[String] = null
  def hasNext: Boolean = {
    if (_next == null) {
      _next = reader !? "next" match {
        case line: String => Some(line)
        case null => { finished = true; Option.empty[String] }
      }
      writer ! "next"
    }
    _next match {
      case Some(_) => true
      case None    => false
    }
  }

  def next: Graph = {
    val res = if (hasNext) { _next.get } else { throw new RuntimeException("no more element") }
    _next = null
    Graphy.fromString(res)
  }

}
